discovery-infra/tests/base_test.py

```diff
@@ -73,37 +83,47 @@ def cluster_configuration(self) -> ClusterConfig:
         return ClusterConfig()
 
     @pytest.fixture
-    def controller(self, cluster_configuration: ClusterConfig,
-                   prepare_controller_configuration: BaseNodeConfig) -> NodeController:
+    def post_cluster_configuration(self, request: pytest.FixtureRequest,
+                                   cluster_configuration: ClusterConfig) -> ClusterConfig:
+        """
+        Allows the test to modify the cluster configuration by registering a custom fixture.
+        To register the custom fixture you have to mark the test with "post_cluster_configuration" marker.
+        @pytest.mark.post_cluster_configuration(FIXTURE1_NAME.__name__, FIXTURE2_NAME.__name__)
+        """
+        yield from utils.run_marked_fixture(cluster_configuration, "post_cluster_configuration", request)
+
+    @pytest.fixture
+    def controller(self, post_cluster_configuration: ClusterConfig,
+                   post_controller_configuration: BaseNodeConfig) -> NodeController:
 
-        if cluster_configuration.platform == consts.Platforms.VSPHERE:
-            return VSphereController(prepare_controller_configuration, cluster_configuration)
+        if post_cluster_configuration.platform == consts.Platforms.VSPHERE:
+            return VSphereController(post_controller_configuration, post_cluster_configuration)
 
-        return TerraformController(prepare_controller_configuration, cluster_config=cluster_configuration)
+        return TerraformController(post_controller_configuration, cluster_config=post_cluster_configuration)
 
     @pytest.fixture
     def nodes(self, controller: NodeController) -> Nodes:
         return Nodes(controller)
 
     @pytest.fixture
-    def prepare_nodes(self, nodes: Nodes, cluster_configuration: ClusterConfig) -> Nodes:
+    def prepare_nodes(self, nodes: Nodes, post_cluster_configuration: ClusterConfig) -> Nodes:
         try:
             nodes.prepare_nodes()
             yield nodes
         finally:
             if global_variables.test_teardown:
                 logging.info('--- TEARDOWN --- node controller\n')
                 nodes.destroy_all_nodes()
-                logging.info(f'--- TEARDOWN --- deleting iso file from: {cluster_configuration.iso_download_path}\n')
-                infra_utils.run_command(f"rm -f {cluster_configuration.iso_download_path}", shell=True)
+                logging.info(f'--- TEARDOWN --- deleting iso file from: {post_cluster_configuration.iso_download_path}\n')
+                infra_utils.run_command(f"rm -f {post_cluster_configuration.iso_download_path}", shell=True)
 
     @pytest.fixture
-    def prepare_network(self, prepare_nodes: Nodes, prepare_controller_configuration: BaseNodeConfig) -> Nodes:
+    def prepare_network(self, prepare_nodes: Nodes, post_controller_configuration: BaseNodeConfig) -> Nodes:
```

Comment:
can we rename to something like _get_nodes_?

Reply:
changed it to controller_configuration because this is the final result. and first fixture which creates it to new_controller_configuration